cmd/pod-scaler/storage.go

```diff
@@ -0,0 +1,141 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"time"
+
+	"cloud.google.com/go/storage"
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/interrupts"
+)
+
+// Cache closes over how we interact with cached data
+type Cache interface {
+	Loader
+	Storer
+	AttributeResolver
+}
+
+// Loader closes over how we load cached data
+type Loader interface {
+	Load(name string) (io.ReadCloser, error)
+}
+
+// Storer closes over how we store cached data
+type Storer interface {
+	Store(name string) (io.WriteCloser, error)
+}
+
+// AttributeResolver closes over how we store cached data
+type AttributeResolver interface {
+	LastUpdated(name string) (time.Time, error)
+}
+
+type bucketCache struct {
+	bucket *storage.BucketHandle
+}
+
+var _ Cache = &bucketCache{}
+
+func (b *bucketCache) Load(name string) (io.ReadCloser, error) {
+	handle := b.bucket.Object(name)
+	return handle.NewReader(interrupts.Context())
+}
+
+func (b *bucketCache) Store(name string) (io.WriteCloser, error) {
+	handle := b.bucket.Object(name)
+	return handle.NewWriter(context.Background()), nil
+}
+
+func (b *bucketCache) LastUpdated(name string) (time.Time, error) {
+	handle := b.bucket.Object(name)
+	attrs, err := handle.Attrs(interrupts.Context())
+	if err != nil {
+		return time.Time{}, fmt.Errorf("could not query cache for attributes: %w", err)
+	}
+	return attrs.Updated, nil
+}
+
+type localCache struct {
+	dir string
+}
+
+var _ Cache = &localCache{}
+
+func (l *localCache) Load(name string) (io.ReadCloser, error) {
+	return os.Open(path.Join(l.dir, name))
+}
+
+func (l *localCache) Store(name string) (io.WriteCloser, error) {
+	return os.Create(path.Join(l.dir, name))
+}
+
+func (l *localCache) LastUpdated(name string) (time.Time, error) {
+	info, err := os.Stat(path.Join(l.dir, name))
+	if err != nil {
+		return time.Time{}, fmt.Errorf("could not query cache for attributes: %w", err)
+	}
+	return info.ModTime(), nil
+}
+
+// LoadCache loads cached query data from the canonical location in GCS.
+func LoadCache(loader Loader, metricName string, logger *logrus.Entry) (*CachedQuery, error) {
```

Comment:
Is there a need for this to be exported?

Reply:
Kind of - this is all in an unimportable `main` right now anyway, but I figured we could move things out into packages before merge? WDYT?